# Esame Laboratorio di Programmazione II - 22/07/2025


Scrivete chiaramente sul notebook il vostro nome e matricola e rinominate il file che consegnate come `matricola.ipynb` (la matricola è SMnumeri o ECnumeri).

Per ogni funzione e metodo che richiede un campionamento rispetto ad una distribuzione settate il seed a 0: ``np.random.seed(0)``.

Stampate il risultato delle domande e consegnate il compito eseguito, quindi per cui ogni cella ha il corrispondente output.
Quindi ad esempio

e.g., 
NON scrivete: 
```python 
    lista=np.array([1,2,3])
```
ma 
```python
    lista=np.array([1,2,3]) 
    print('lista =', lista)   
```
Attenzione!Se scrivete:
```python
    lista=np.array([1,2,3])
    lista
```
``lista`` sarà l'unico ouput che si vede di quella cella

NB: Potete usare anche più di una cella python per domanda e se l'output è una matrice molto grande non dovete stamparla.

Sarà valutata anche la presentazione, ad esempio un plot senza etichette sugli assi o illeggibili sarà valutato meno bene di uno con etichette chiare.

Cercate di commmentare il compito di modo che si capisca cosa avete fatto.

Controllate attentamente di avere consegnato il file giusto.

## Esercizio 1

Crea un vettore NumPy con le seguenti spese annuali per l’auto (in euro) sostenute da 7 persone:
3200, 2750, 4100, 3600, 2950, 5100, 4400

1. Calcola la spesa totale complessiva sostenuta dal gruppo.

2. Le persone con spese minori di 3000 € ottengono una detrazione fiscale del 15%. Applica la modifica al vettore.

3. Quanto ha risparmiato il gruppo dopo la detrazione fiscale?


In [ ]:
import numpy as np
speseAnnuali=np.array([3200, 2750, 4100, 3600, 2950, 5100, 4400], dtype=float)
somma=np.sum(speseAnnuali)
speseAnnuali[speseAnnuali<3000]=speseAnnuali[speseAnnuali<3000]*0.85
diff=somma-np.sum(speseAnnuali)
print(somma)
print(speseAnnuali)
print(diff)


26100.0
[3200.  2337.5 4100.  3600.  2507.5 5100.  4400. ]
855.0


## Esercizio 2
1. Scrivi una funzione che simula il lancio di un dado per N volte e restituisce un array con il numero di volte che è uscita ciascuna faccia (da 1 a 6).
2. Calcolare la frequenza % con la dimensione del campione (numero di lanci) 100 volte da 10 a 5000.  Ottieni così 100 array di frequenze.
3. Fai il plot dell’andamento della frequenza media delle facce 1, 3 e 6 in funzione della dimensione del campione.



In [52]:

def lanci(N):
    array=np.array(np.random.choice(np.arange(1,7),N,replace=True))
    return np.bincount(array, minlength=7)[1:]
lanci(100)
volte=np.linspace(10,5000,100,dtype=int)
frequenze=np.array([lanci(x)/x for x in volte] )
frequenze


array([[0.2       , 0.1       , 0.4       , 0.2       , 0.        ,
        0.1       ],
       [0.15      , 0.06666667, 0.13333333, 0.26666667, 0.13333333,
        0.25      ],
       [0.13636364, 0.19090909, 0.11818182, 0.14545455, 0.17272727,
        0.23636364],
       [0.19875776, 0.11180124, 0.16149068, 0.2173913 , 0.16770186,
        0.14285714],
       [0.16113744, 0.13744076, 0.20379147, 0.17535545, 0.18009479,
        0.14218009],
       [0.19465649, 0.15648855, 0.17938931, 0.16030534, 0.14885496,
        0.16030534],
       [0.14423077, 0.16025641, 0.13141026, 0.16346154, 0.21474359,
        0.18589744],
       [0.18232044, 0.21270718, 0.13812155, 0.17127072, 0.13812155,
        0.15745856],
       [0.1622276 , 0.17917676, 0.18159806, 0.16464891, 0.13559322,
        0.17675545],
       [0.18142549, 0.19006479, 0.174946  , 0.15334773, 0.13390929,
        0.1663067 ],
       [0.19066148, 0.15564202, 0.14202335, 0.17315175, 0.17315175,
        0.16536965],
       [0.16134752, 0

## Esercizio 3

1. Genera una matrice NumPy di dimensioni 7 x 7, con valori estratti da una distribuzione normale con media = 0, ed std = 5.

2. Per ogni riga, identifica il valore più lontano da zero (cioè con valore assoluto massimo). Crea un array con questi 7 valori (uno per riga).

3. Calcola quante di queste righe hanno il valore più lontano da zero positivo?

In [70]:
matrice=5 * np.random.randn(7,7)
print(matrice)
valLontano=np.argmax(abs(matrice), axis=1)
valori=matrice[np.arange(7),valLontano]
somma=np.sum(valori>0)
print(valori)
somma

[[ 1.26724181  9.91718706  4.96658837  1.59102926 -1.5242569   3.20893778
  -3.86913493]
 [-3.98982367  2.84099948 -3.29020994  6.76330957 -7.67544439  4.41980952
   6.20783892]
 [-3.13408381 -1.78241844  0.95579906 -2.63892959  1.89676352 -7.6267165
   3.47062523]
 [ 3.99170118  9.32831514 -3.25817427  2.20825732  4.38928749  4.16135093
  -1.08688049]
 [ 5.040031   -0.20824377 -3.7479014  -1.61431982  2.1195268  -7.26796211
  -0.74136501]
 [-6.63951792  2.49835381 -6.3614751  -3.09799318  1.19024986 -1.35630815
  -5.610381  ]
 [ 3.24961301  4.53756943  1.6850331   2.73103864  6.96909438 -7.41166622
  -8.49959459]]
[ 9.91718706 -7.67544439 -7.6267165   9.32831514 -7.26796211 -6.63951792
 -8.49959459]


np.int64(2)

## Esercizio 4 

Utilizza il dataset cereal.csv.

Contiene dati nutrizionali (proteine, fibre, calorie, zuccheri, ecc.) per circa 70 cereali.


1.  Aggiungi una colonna chiamata `nutri_score`, definita come:
$$
\texttt{nutri\_score} = \frac{\texttt{protein} + \texttt{fiber}}{\texttt{calories}}
$$

Indica quanti nutrienti positivi ci sono per ogni caloria: più è alto, più il cereale è "nutriente a parità di energia.

2. Trova i **5 cereali con il `nutri_score` più alto** e stampali con `name` e `nutri_score`.

3. Filtra i cereali che soddisfano entrambi i criteri:

   - meno di 120 calorie  
   - più di 2 grammi di fibra 


4. **Crea un grafico a barre** dei **10 cereali con il `nutri_score` più alto**, usando `name` sull’asse x.

5. Crea uno **scatter plot** (grafico a punti) che visualizzi:

   - `calories` sull’asse x  
   - `nutri_score` sull’asse y
   - `protein` nel colore dei punti


## Esercizio 5 

Utilizza il dataset `imdb_top_1000.csv`, che contiene informazioni su 1000 film recensiti su IMDB.

1. Quanti **valori mancanti** ci sono per colonna? Elimina le righe con i valori mancanti.

2. Elimina la colonna `Overview`. Aggiungi una colonna (`main_genre`) con solo il primo genere della colonna `genre`. Quali sono i 5 generi più frequenti?

3. Che attore/attrice è stato/a più volte la `Star1` e quello/a che ha fatto più film.

4. Calcola l'IMDB_rating medio e il Gross medio di ogni genere. Fai uno scatter plot dell'IMDB_rating medio e il Gross medio di ogni genere. Riesci a mettere anche l'ettichetta del genere?

5. Visualizza la distribuzione del IMDB_rating  e la distribuzione del `Gross` dei primi 5 main_genre. (Hint fai due violin plot).




## Esercizio 6

Considera i seguenti dati, che :

`x = np.linspace(0, 10, 100)`  
`y = -3.5 * x + 2 + np.random.normal(0, 10, 100)`

Cerca di riprodurre la retta fittando i parametri della funzione y = ax + b.

1. Stima i parametri della retta.
2. Calcola l’intervallo di confidenza al 95% dei parametri a e b.
N.B il secondo output del metodo curve_fit è proprio la matrice di covarianza: `params, cov = curve_fit(retta, x, y)`. 
L'**errore standard** di ciascun parametro si calcola come:

$$
\text{errore\_standard}_i = \sqrt{\text{cov}(i, i)}
$$

dove $\text{cov}(i, i)$ è l'elemento diagonale $i$-esimo della **matrice di covarianza**.

L’**intervallo di confidenza al 95%** si approssima con:

$$
\text{parametro}_i \pm 1.96 \times \text{errore\_standard}_i
$$
3. Plotta i punti originali e la retta ottenuta con anche l'intervallo di confidenza